src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

```diff
@@ -348,10 +352,15 @@ void Complete()
                         {
                             result = System.Array.Empty<object>();
                         }
+                        else if (IsAssignableToGenericTaskType(task_type))
+                        {
+                            result = task_type.GetMethod("get_Result")?.Invoke(task, null);
```

Comment:
Would it make sense to propose addition to reflection APIs that would do: "given a MethodInfo for a method on a generic definition of the type, find the MethodInfo on this instantiated type"?

(We could then cache the MethodInfo for the generic definition on the side and all of this would be 100% trim safe without annotations.)

Having to do name-binding here is silly (we could implement the former using a quick token lookup), plus this will do funny things when some jokers implement their own tasks:

```csharp
class MyTask : Task<string>
{
    public MyTask(Func<string> function)
        : base(function) { }

    public new string Result => "Haha fooled ya";
}
```

Reply:
Actually, the joker case is going to not to the same thing at runtime after trimming because the DynamicDependency is in fact not equivalent.

Reply:
I thought about the above case, but figured it was too far fetched. If someone returned ‘MyTask’ as a Task{string}, getting the Result wouldn’t be correct. And there is no real way to set the base Result property so both properties return the same value.

To solve that scenario, we could walk the inheritance hierarchy until we found Task{TResult} and then grabbed the Result property directly from that type.

> Would it make sense to propose addition to reflection APIs that would do: "given a MethodInfo for a method on a generic definition of the type, find the MethodInfo on this instantiated type"?

I think this is a good idea. I was trying to figure out how to do this with the current APIs (like calling MakeGenericMethod) but was unable. I’ll write it up today.

Reply:
> I’ll write it up today.

Thank you! I think it will be a useful addition - I've seen people use reflection similar to what's here and it suffers the same correctness issues as the existing code (matches any `get_Result` method in any of the derived classes) - and it is not trimming friendly, even though it could be.

I dislike DynamicDependency+Suppress because it results in missed corner cases and bugs. Different behavior between trimmed/untrimmed (without a linker warning) should always be considered a bug.

Reply:
> I dislike DynamicDependency+Suppress because it results in missed corner cases and bugs. Different behavior between trimmed/untrimmed (without a linker warning) should always be considered a bug.

Without the above API addition, is there currently a better way than `DynamicDependency+Suppress`?

Reply:
I couldn't find a trimming-safe way. MemberInfo.HasSameDefinitionAs could be used to implement poor man's version of the "give me this method on the instantiated version of the type", but that approach requires iterating over all methods - not sure if the perf would be acceptable and it still requires a suppression (we would be able to get by without the DynamicDependency though).

Reply:
Logged https://github.com/dotnet/runtime/issues/45771 for this API. Thanks for the suggestion.